pkg/controller/cronjob/cronjob_controller.go

```diff
@@ -119,10 +120,92 @@ func (jm *CronJobController) SyncAll() {
 	glog.V(4).Infof("Found %d groups", len(jobsBySj))
 
 	for _, sj := range sjs {
-		SyncOne(sj, jobsBySj[sj.UID], time.Now(), jm.jobControl, jm.sjControl, jm.podControl, jm.recorder)
+		jobsRemaining := CleanupFinishedJobs(sj, jobsBySj[sj.UID], jm.jobControl, jm.recorder)
+		SyncOne(sj, jobsRemaining, time.Now(), jm.jobControl, jm.sjControl, jm.podControl, jm.recorder)
```

Comment:
In principle, I am not sure I can avoid to remove from active list though -- if the history limit is set to 0 (or to a low value, or there are many cronjobs maybe), cleanup may remove a job that finished in between syncing the cronjob and cleaning it up, which would still lead to an undesirable warning next time the cronjob is sync-ed.

Since I need to reap the pods manually, how about taking the controller's reaping code after
```.go
	if sj.Spec.ConcurrencyPolicy == batch.ReplaceConcurrent {
		for _, j := range sj.Status.Active {
```
out in a utility function, and also using that? This removes the pods, the job, and the reference in active list.